File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def helper(root, max, min):
            if(not root):
                return True
            
            if(root.val >= max):
                return False
            if(root.val <= min):
                return False
            return helper(root.left,root.val,min) and helper(root.right,max,root.val)


        return helper(root,inf,-inf)

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root, max = Infinity, min = -Infinity) {
    if(!root) return true;
    if(root.val >= max || root.val <= min) return false;
    // let newMax = Math.min(max,root.val);
    // let newMin = Math.max(min,root.val);
    return(isValidBST(root.left,root.val,min) && isValidBST(root.right,max,root.val));
};